Feat/generate challenge

ssardana96 opened the pull request:
## Changes

## Fixes


## Tests 
Tasks to complete before merging PR:
- [ ]  Ensure system tests are passing. If not [Run them manually](https://github.com/0chain/0chain/actions/workflows/system_tests.yml) to check for any regressions :clipboard:
- [ ]  Do any new system tests need added to test this change? do any existing system tests need updated? If so create a PR at [0chain/system_test](https://github.com/0chain/system_test)
- [ ]  Merge your system tests PR to master AFTER merging this PR

## Associated PRs (Link as appropriate):
- blobber:
- gosdk:
- system_test: https://github.com/0chain/system_test/pull/224
- zboxcli:
- zwalletcli:
- Other: ...

ssardana08 committed:
- blobber challenge partition added
- partition added to commit challenge
- new methods added to partition

ssardana08 committed:
- UpdateItem added

ssardana08 committed:
- partition save added

ssardana08 committed:
- challenge partition write updated

ssardana08 committed:
- correct partition functions

ssardana08 committed:
- generateChallenge compiled

ssardana08 committed:
- added allocation challenge
- cleaning blobber challenge

ssardana08 committed:
- blobber challenge updated

ssardana08 committed:
- blobberid added to storage challenge
- storage challenge to be stored in MPT

ssardana08 committed:
- emitaddoverwrite blobber moved down

ssardana08 committed:
- challenge getting stored to MPT
- fix completeChallengeForBlobber

ssardana08 committed:
- removed challenges concept and added challendge id only to blobber challenge
- removed validators

ssardana08 committed:
- canceledPassRates refactored to have only 1 MPT access

ssardana08 committed:
- complete challenge for blobber fixed

ssardana08 committed:
- fix benchmarks

ssardana08 committed:
- unit tests fixed

ssardana08 committed:
- benchmark setup

ssardana08 committed:
- removed adding of challenges

ssardana08 committed:
- added allocation id to challenge benchmark

ssardana08 committed:
- adding challenge to test benchmark

ssardana08 committed:
- adding storage challenge to mpt in benchmark

ssardana08 committed:
- revert challenge storing to mpt logic

ssardana08 committed:
- revert challenge storing to mpt logic

dabasov commented on code/go/0chain.net/smartcontract/storagesc/alloation2_test.go:
```diff
@@ -363,17 +363,30 @@ func testCancelAllocation(
+		//		AllocationID: sAllocation.ID,
+		//		Created:      created,
+		//	})
+		//}
```
can we delete this dead code?

ssardana08 committed:
- correcting benchmark
- partition add partition corrected

Sriep commented on code/go/0chain.net/smartcontract/storagesc/challenge.go:
```diff
@@ -581,27 +689,21 @@ func (sc *StorageSmartContract) generateChallenges(t *transaction.Transaction,
 			"error getting the validators list: %v", err)
 	}
 
-	var all *Allocations
```
At statt of function.
```go
	var tp = time.Now()
	defer sc.addGenerateChallengesStat(tp, &err)
```
Could we move this out of the smart-contract. It should not have access to non-deterministic data.

ssardana96 replied:
done

Sriep commented on code/go/0chain.net/smartcontract/storagesc/challenge.go:
```diff
@@ -581,27 +689,21 @@ func (sc *StorageSmartContract) generateChallenges(t *transaction.Transaction,
-
-	var selectAlloc = func(i int) (alloc *StorageAllocation, err error) {
-		alloc, err = sc.getAllocation(all.List[i], balances)
+	var selectAlloc = func(allocID string) (alloc *StorageAllocation, err error) {
```
Why use an Anonymous function to validate an allocation, rather than a regular one?

Sriep commented on code/go/0chain.net/smartcontract/partitions/blobber_challenge_list.go:
```diff
@@ -0,0 +1,172 @@
+//------------------------------------------------------------------------------
+
+type BlobberChallengeNode struct {
+	ID           string `json:"id"`
```
Maybe `BlobberID` rather than `ID`. This is not the id of the `BlobberChallengeNode` but the id of some blobber somewhere else.

Sriep commented on code/go/0chain.net/smartcontract/storagesc/challenge.go:
```diff
@@ -617,17 +719,35 @@ func (sc *StorageSmartContract) generateChallenges(t *transaction.Transaction,
+
+		randomIndex := rand.Intn(len(bcPartition))
+		bcItem := bcPartition[randomIndex]
+		val, ok := bcItem.(*partitions.BlobberChallengeNode)
```
Ugh! could we hide this bit of code in the Partitions package. Maybe add a global  `GetRandomAllocationBlobberPair` `functions to blobber_callenge_list.go` for the previous twenty or so lines of code.

Sriep commented on code/go/0chain.net/smartcontract/storagesc/challenge.go:
```diff
@@ -617,17 +719,35 @@ func (sc *StorageSmartContract) generateChallenges(t *transaction.Transaction,
 
 	for i := int64(0); i < numChallenges; i++ {
 
-		// looking for allocation with NumWrites > 0
```
Could we split this `for` loop into two `for` loops? 
1. Call all the `GetTrieNode` methods. Make this a parallel `for` loop using wait groups or such.
2. Use a synchronous `for` loop for all the `InsertTrieNode` calls.

We are getting a few hundred items from the MPT, this could save a significant amount of time.

Probably best to do most of the work in the first for loop. Create a slice of `allocChallengeObj`, `storageChallenge`, `blobberChallengeObj` and `alloc` objects to be inserted into the MPT in the second for loop.

Sriep commented on code/go/0chain.net/smartcontract/storagesc/challenge.go:
```diff
@@ -667,48 +777,43 @@ func (sc *StorageSmartContract) generateChallenges(t *transaction.Transaction,
+			break
 		}
 	}
+	if blobber == nil {
```
blobber is never nil as set to `&StorageNode{}` above. Check `blobber.ID == ""` or similar instead.

Sriep commented on code/go/0chain.net/smartcontract/storagesc/challenge.go:
```diff
@@ -722,39 +827,77 @@ func (sc *StorageSmartContract) addChallenge(alloc *StorageAllocation,
 	}
 
-	balances.InsertTrieNode(blobberChallengeObj.GetKey(sc.ID), blobberChallengeObj)
+	_, err = balances.InsertTrieNode(allocChallengeObj.GetKey(sc.ID), allocChallengeObj)
```
Separate out the following block of code with the  `InsertTrieNode` calls, and put in a separate synchronous second for loop.

Sriep commented on code/go/0chain.net/smartcontract/storagesc/models.go:
```diff
@@ -117,6 +121,61 @@ func (sn *BlobberChallenge) GetHashBytes() []byte {
+
+type AllocationChallenge struct {
+	AllocationID             string                       `json:"allocation_id"`
+	Challenges               []*StorageChallenge          `json:"challenges"`
```
This slice of `StorageChallenge` objects is too big. I think we need to come up with a way of avoiding needing it. In the meanwhile, the StoreaChallenger object only needs to know its Id and Created.

Sriep commented on code/go/0chain.net/smartcontract/storagesc/models.go:
```diff
@@ -144,19 +204,47 @@ func (sn *BlobberChallenge) addChallenge(challenge *StorageChallenge) bool {
-	RandomNumber   int64              `json:"seed"`
-	AllocationID   string             `json:"allocation_id"`
-	Blobber        *StorageNode       `json:"blobber"`
+	Created common.Timestamp `json:"created"` // needed
```
Why do we need anything other than ID and Created here *(for `canceledPassRates`)? 

This needs to be as slim as possible or ideally removed altogether. Most of this data is to be stored in the event database not the MPT. Only keep what is absolutely essential in the MPT.

Sriep commented on code/go/0chain.net/smartcontract/storagesc/challenge.go:
```diff
@@ -722,39 +827,77 @@ func (sc *StorageSmartContract) addChallenge(alloc *StorageAllocation,
+			"error storing alloc challenge: "+err.Error())
+	}
+
+	_, err = balances.InsertTrieNode(storageChallenge.GetKey(sc.ID), &storageChallenge)
```
The storageChallenge object should be emitted to the event database and stored in the Challenge table, not stored in the MPT. For example see https://github.com/0chain/0chain/pull/679 
[balances.EmitEvent(event.TypeStats, event.TagAddChallenge, challengeID, string(data))](https://github.com/0chain/0chain/blob/challenge-events/code/go/0chain.net/smartcontract/storagesc/challenge.go#L809)

This way we can remove everything from StorageChallenge except what we actually need for smart contracts. Ideally, it can be replaced altogether with summary stats.

Sriep commented on code/go/0chain.net/smartcontract/storagesc/challenge.go:
```diff
@@ -722,39 +827,77 @@ func (sc *StorageSmartContract) addChallenge(alloc *StorageAllocation,
+	}
 	//Logger.Info("Adding a new challenge", zap.Any("blobberChallengeObj", blobberChallengeObj), zap.Any("challenge", storageChallenge.ID))
 	challengeBytes, err := json.Marshal(storageChallenge)
 	sc.newChallenge(balances, storageChallenge.Created)
```
`sc.newChallenge(balances, storageChallenge.Created)` I am not convinced we need this object, but it does not need to be updated once per loop, we can do this once at the end outside the for loops.

Sriep reviewed (changes requested):
We need a new Challenge table https://github.com/0chain/0chain/issues/769 for the event database if this PR is to work  For example https://github.com/0chain/0chain/pull/679. This can either be done first or included as part of this PR?

ssardana08 committed:
- added blobberchallengeallocation partition

ssardana08 committed:
- fixed benchmark setup allocation id

ssardana08 committed:
- benchmark setup fixed

ssardana08 committed:
- fixed blobberChallengeItemList add

ssardana08 committed:
- commit connection fixed

ssardana08 committed:
- async generate challenge

ssardana08 committed:
- few neats

peterlimg commented on code/go/0chain.net/smartcontract/storagesc/blobber.go:
```diff
@@ -1,19 +1,20 @@
 
 import (
+	"0chain.net/core/logging"
+	"0chain.net/smartcontract/partitions"
```
run goimports to format

ssardana96 replied:
running goimport formatted imports
but how did you get to know we should run it?

peterlimg replied:
The packages sequence is different if not formatted.

peterlimg commented on code/go/0chain.net/smartcontract/partitions/blobber_challenge.go:
```diff
@@ -0,0 +1,166 @@
+
+func (il *blobberChallengeItemList) add(it PartitionItem) error {
+
+	il.Items = append(il.Items, BlobberChallengeNode{
```
Check duplicate and return error like bbr

peterlimg commented on code/go/0chain.net/smartcontract/storagesc/challenge.go:
```diff
@@ -28,23 +28,87 @@ import (
 	idx := 0
-	if found && idx >= 0 && idx < len(blobberChallengeObj.Challenges) {
-		blobberChallengeObj.Challenges = append(blobberChallengeObj.Challenges[:idx], blobberChallengeObj.Challenges[idx+1:]...)
+	if found && idx < len(blobberChallengeObj.ChallengeIDs) {
```
If found is true, the idx is always 0 I think, so perhaps we can use ChallengeIDs = ChallengeIDs[1:] to remove it and make this code cleaner

peterlimg commented on code/go/0chain.net/smartcontract/storagesc/challenge.go:
```diff
@@ -513,6 +640,104 @@ func (sc *StorageSmartContract) addGenerateChallengesStat(tp time.Time,
+		Logger.Error("client state has invalid allocations",
+			zap.Any("selected_allocation", allocID))
+		return nil, common.NewErrorf("invalid_allocation",
+			"client state has invalid allocations")
```
Use switch case to make this error handling easier to read, e.g:
switch err {
case nil:
case util.ErrValueNotPresent:
default:

peterlimg commented on code/go/0chain.net/smartcontract/storagesc/challenge.go:
```diff
@@ -513,6 +640,104 @@ func (sc *StorageSmartContract) addGenerateChallengesStat(tp time.Time,
 	}
 }
 
+func (sc *StorageSmartContract) selectAllocationForChallenge(
```
function name as getAllocationForChallenge would be more reasonable, as we don't have any 'select' action here, we just get it from mpt and validate it.

peterlimg commented on code/go/0chain.net/smartcontract/storagesc/challenge.go:
```diff
@@ -513,6 +640,104 @@ func (sc *StorageSmartContract) addGenerateChallengesStat(tp time.Time,
+				"error getting random slice from blobber challenge partition")
+		}
+
+		randomIndex := rand.Intn(len(bcPartition))
```
The randomIndex will make the state inconsistent when the block is execute on different miners/sharders, and lead  to state hash mismatch error.  We need to run it with the same random seed.

ssardana96 replied:
peter can you check this now.

peterlimg commented on code/go/0chain.net/smartcontract/storagesc/challenge.go:
```diff
@@ -513,6 +640,104 @@ func (sc *StorageSmartContract) addGenerateChallengesStat(tp time.Time,
+			errChan <- common.NewError("generate_challenges",
+				"error getting random slice from blobber challenge allocation partition")
+		}
+		randomIndex = rand.Intn(len(bcAllocPartition))
```
same issue as above

ssardana96 replied:
can u check this now

peterlimg commented on code/go/0chain.net/smartcontract/storagesc/challenge.go:
```diff
@@ -28,23 +28,87 @@ import (
+}
+
+func getBlobbersChallengeAllocationList(blobberID string, balances c_state.StateContextI) (partitions.RandPartition, error) {
+	all, err := partitions.GetRandomSelector(getBlobberChallengeAllocationKey(blobberID), balances)
```
When will this partition be updated?

ssardana96 replied:
in commitBlobberConnection

peterlimg commented on code/go/0chain.net/smartcontract/partitions/itemlist.go:
```diff
@@ -50,9 +51,10 @@ func (il *itemList) get(key datastore.Key, balances state.StateContextI) error {
 
-func (il *itemList) add(it PartitionItem) {
+func (il *itemList) add(it PartitionItem) error {
 	il.Items = append(il.Items, StringItem{it.Name()})
```
Check duplicate item

peterlimg commented on code/go/0chain.net/smartcontract/storagesc/models.go:
```diff
@@ -93,10 +99,12 @@ type ChallengeResponse struct {
-	LatestCompletedChallenge *StorageChallenge            `json:"lastest_completed_challenge"`
+	BlobberID string `json:"blobber_id"`
+	//Challenges               []*StorageChallenge          `json:"challenges"`                  //to be removed
+	//ChallengeMap             map[string]*StorageChallenge `json:"-"`                           // to be removed
```
Remove if these commented code is not used anymore

peterlimg commented on docker.local/bin/unit_test.sh:
```diff
@@ -43,5 +43,5 @@ if [[ -n "$PACKAGE" ]]; then
     # Run all tests.
     echo "Running general unit tests:"
-    docker run "$INTERACTIVE" zchain_unit_test sh -c "cd 0chain.net; go test -tags bn256 -cover ./..."
+    docker run "$INTERACTIVE" zchain_unit_test sh -c "cd 0chain.neti/smartcontract/storagesc; go test -tags bn256 -cover ./..."
```
Seems like local changes for testing

Sriep commented on code/go/0chain.net/smartcontract/storagesc/alloation2_test.go:
```diff
@@ -363,17 +363,30 @@ func testCancelAllocation(
+		require.NoError(t, err)
 	}
 
 	allAllocationsBefore, err := ssc.getAllAllocationsList(ctx)
```
The only user for the `ALL_ALLOCATIONS_KEY` and the all allcoatinos list is in genearteChallengers. As we are not going to use it anymore all references to `ALL_ALLOCATIONS_KEY` can be removed. 

This will save a lot of time across several endpoints.

ssardana08 committed:
- blobber id name fix
- stat from generate challenge removed

ssardana08 committed:
- insert and gettrieNode separated out

ssardana08 committed:
- code rebased

ssardana08 committed:
- benchmark add challenge fixed

ssardana08 committed:
- test fixed

ssardana08 committed:
- test fixed

ssardana08 committed:
- challenge error handling

ssardana08 committed:
- tests fixed

ssardana08 committed:
- tests fixed

ssardana08 committed:
- add challenge test fixed with validators

ssardana08 committed:
- removed storagestats

ssardana08 committed:
- blobber challenge add check added

ssardana08 committed:
- remove commented code
- remove getAllAllocationsList
- completeChallengeForBlobber neats
- getAllocationsForChallenge err handled better
- change rand seed
- add method improved for blobberChallenfe and itemlist

ssardana08 committed:
- storageChallenge size reduced

ssardana08 committed:
- combined error to result

ssardana08 committed:
generatechallenge

ssardana08 committed:
- canceledPassRates logical errors

service-0chain commented:
Manual system tests [failure] with the following config
| config  | value |
| ------------- | ------------- |
| tests skipped  | FALSE  |
| system-tests  | feat/generate_challenge  |
| gosdk  | NOT OVERRIDDEN  |
| zboxcli  | staging  |
| zwalletcli  | staging  |
| 0chain  | feat/generate_challenge  |
| blobber  | staging  |
| authorizer  | staging  |
| 0box  | staging  |
| 0block  | staging  |
| 0dns  | staging  |
| explorer  | staging  |
| 0proxy  | staging  |
| 0search  | staging  |

Sriep commented on code/go/0chain.net/smartcontract/storagesc/challenge.go:
```diff
@@ -498,43 +591,290 @@ func (sc *StorageSmartContract) addGenerateChallengesStat(tp time.Time,
+	)
+	defer close(data)
+
+	go sc.asyncGenerateChallenges(blobberChallengeList, validators, r, t.CreationDate, data, output, balances)
```
Why are we running this  in a goroutine? We are only generating one challenge so there is no need for this.

ssardana96 replied:
this has channels. hence goroutine

Sriep commented on code/go/0chain.net/smartcontract/storagesc/challenge.go:
```diff
@@ -498,43 +591,290 @@ func (sc *StorageSmartContract) addGenerateChallengesStat(tp time.Time,
+
+	defer close(result)
+
+	for d := range data {
```
When only creating one challenge we want to create it synchronously. In which case, there will be only one entry in data.

I suggest putting the content of this for loop in a separate function that can be called directly from generateChallenge instead of `asyncGenerateChallenges` and also from for loop.

Sriep commented on code/go/0chain.net/smartcontract/storagesc/challenge.go:
```diff
@@ -498,43 +591,290 @@ func (sc *StorageSmartContract) addGenerateChallengesStat(tp time.Time,
+
+	var (
+		data   = make(chan challengeInput)
+		output = make(chan challengeOutput)
```
We are only creating one challenge so there is no need to use channels.

Sriep commented on code/go/0chain.net/smartcontract/storagesc/challenge.go:
```diff
@@ -498,43 +591,290 @@ func (sc *StorageSmartContract) addGenerateChallengesStat(tp time.Time,
+	data <- challengeInput{
+		cr:          cr,
+		t:           t,
+		challengeID: challengeID,
```
All this code involving channels can be removed.

ssardana08 committed:
- generateChallenge refactored

ssardana96 committed:
Merge branch 'staging' into feat/generate_challenge

ssardana08 committed:
- merge fixes

ssardana08 committed:
- lint fixes

ssardana08 committed:
- lint fixes

ssardana08 committed:
- allocChallenges added in benchmark

ssardana08 committed:
- benchmark populating challenge partition

Sriep commented on code/go/0chain.net/smartcontract/storagesc/benchmark_tests.go:
```diff
@@ -748,7 +748,7 @@ func BenchmarkTests(
 				challengesEnabled := viper.GetBool(bk.StorageChallengeEnabled)
 				if challengesEnabled {
-					err := ssc.generateChallenges(txn, balances.GetBlock(), nil, balances)
+					err := ssc.generateChallenge(txn, balances.GetBlock(), nil, balances)
```
Change the name as well, ~s~ 
```go
			name: "storage.generate_challenge",
```

ssardana08 committed:
- error added

ssardana08 committed:
merge fixes

ssardana08 committed:
- generate_challenge and generate_challenges added

ssardana08 committed:
- lint fixes

ssardana08 committed:
- logs added for debugging st

ssardana08 committed:
- logs added for debugging st

ssardana08 committed:
- blobber challenge partition location added

service-0chain commented:
Manual system tests [success] with the following config
| config  | value |
| ------------- | ------------- |
| tests skipped  | FALSE  |
| system-tests  | feat/generate_challenge  |
| gosdk  | NOT OVERRIDDEN  |
| zboxcli  | staging  |
| zwalletcli  | staging  |
| 0chain  | feat/generate_challenge  |
| blobber  | staging  |
| authorizer  | staging  |
| 0box  | staging  |
| 0block  | staging  |
| 0dns  | staging  |
| explorer  | staging  |
| 0proxy  | staging  |
| 0search  | staging  |

ssardana96 committed:
Merge branch 'staging' into feat/generate_challenge

ssardana96 committed:
Merge branch 'staging' into feat/generate_challenge

ssardana08 committed:
- merge issues

ssardana96 committed:
Merge branch 'staging' into feat/generate_challenge

ssardana08 committed:
- test fixes

ssardana08 committed:
Merge branch 'feat/generate_challenge' of github.com:0chain/0chain into feat/generate_challenge

ssardana08 committed:
- test fixes

ssardana08 committed:
- generate challenge per block

ssardana08 committed:
- added generate_challenge cost

ssardana96 committed:
Merge branch 'staging' into feat/generate_challenge

Sriep commented on code/go/0chain.net/smartcontract/storagesc/sc.go:
```diff
@@ -217,29 +218,29 @@ func (sc *StorageSmartContract) Execute(t *transaction.Transaction,
-				return "", err
-			}
-		}
+		//challengesEnabled := config.SmartContractConfig.GetBool(
```
remove commented out code, same below.

Sriep commented on code/go/0chain.net/smartcontract/storagesc/challenge.go:
```diff
@@ -489,60 +584,275 @@ func (sc *StorageSmartContract) verifyChallenge(t *transaction.Transaction,
+		return common.NewErrorf("adding_challenge_error",
+			"Error in adding challenge: %v", err)
 	}
+	if tm := sc.SmartContractExecutionStats["challenge_request"]; tm != nil {
```
What is this bit of code doing? I can't see any reason to keep it.

ssardana08 committed:
merge fixes

ssardana08 committed:
- commented code removed
- review neats

ssardana96 committed:
Merge branch 'staging' into feat/generate_challenge

ssardana08 committed:
- blobber rewards key updated

ssardana08 committed:
Merge branch 'feat/generate_challenge' of github.com:0chain/0chain into feat/generate_challenge

peterlimg commented on code/go/0chain.net/smartcontract/storagesc/challenge.go:
```diff
@@ -451,7 +541,7 @@ func (sc *StorageSmartContract) verifyChallenge(t *transaction.Transaction,
-		threshold = len(challReq.Validators) / 2
+		threshold = challReq.TotalValidators / 2
 		pass      = success > threshold ||
 			(success > failure && success+failure < threshold)
```
Mark for later discussion about the `pass` calculation

peterlimg commented on code/go/0chain.net/smartcontract/storagesc/challenge.go:
```diff
@@ -451,7 +541,7 @@ func (sc *StorageSmartContract) verifyChallenge(t *transaction.Transaction,
-		threshold = len(challReq.Validators) / 2
+		threshold = challReq.TotalValidators / 2
 		pass      = success > threshold ||
 			(success > failure && success+failure < threshold)
```
Mark for later  discussion about the `pass` calculation and the following partial success reward

ssardana96 replied:
created an issue for the same: https://github.com/0chain/0chain/issues/1128

peterlimg commented on code/go/0chain.net/smartcontract/storagesc/challenge.go:
```diff
@@ -628,85 +716,278 @@ func (sc *StorageSmartContract) verifyChallenge(t *transaction.Transaction,
-	var tm = sc.SmartContractExecutionStats["generate_challenges"]
-	if tm == nil {
-		return // missing timer (unexpected)
+	if alloc.Expiration < t.CreationDate {
```
we can return error to indicate the allocation expiration error

peterlimg commented on code/go/0chain.net/smartcontract/storagesc/challenge.go:
```diff
@@ -628,85 +716,278 @@ func (sc *StorageSmartContract) verifyChallenge(t *transaction.Transaction,
-	if timer, ok := tm.(metrics.Timer); ok {
-		timer.Update(time.Since(tp))
+	if alloc.Stats == nil {
+		return nil, nil
```
Return error to indicate the empty stats

peterlimg commented on code/go/0chain.net/smartcontract/storagesc/challenge.go:
```diff
@@ -628,85 +716,278 @@ func (sc *StorageSmartContract) verifyChallenge(t *transaction.Transaction,
+	}
+
+	blobberAllocation, ok := alloc.BlobberMap[blobber.ID]
+	if !ok {
```
Move this checking above before the blobber finding to avoid unnecessary for loop.

ssardana96 replied:
this is fixed in: https://github.com/0chain/0chain/pull/1120

ssardana96 replied:
fixed

peterlimg commented on code/go/0chain.net/smartcontract/storagesc/challenge.go:
```diff
@@ -628,85 +716,278 @@ func (sc *StorageSmartContract) verifyChallenge(t *transaction.Transaction,
+	}
+
+	selectedValidators := make([]*ValidationNode, 0)
+	randSlice, err := validators.GetRandomSlice(challRand, balances)
```
rename randSlice to randValidators would make it meaningful and easy to read

peterlimg commented on code/go/0chain.net/smartcontract/storagesc/challenge.go:
```diff
@@ -628,85 +716,278 @@ func (sc *StorageSmartContract) verifyChallenge(t *transaction.Transaction,
+
+	perm := challRand.Perm(len(randSlice))
+	for i := 0; i < minInt(len(randSlice), alloc.DataShards+1); i++ {
+		if randSlice[perm[i]].Name() != blobber.ID {
```
Assign as `randValidator := randSlice[perm[i]]` and use it below would make the code cleaner,

peterlimg commented on code/go/0chain.net/smartcontract/storagesc/challenge.go:
```diff
@@ -628,85 +716,278 @@ func (sc *StorageSmartContract) verifyChallenge(t *transaction.Transaction,
 
-	if numMins == 0 {
-		numMins = 1
+	challengeID := encryption.Hash(hashString + strconv.FormatInt(int64(1), 10))
```
Unnecessary type convert from `int(1)` to string, is it needed if its value is always `1`?

peterlimg commented on code/go/0chain.net/smartcontract/storagesc/challenge.go:
```diff
@@ -628,85 +716,278 @@ func (sc *StorageSmartContract) verifyChallenge(t *transaction.Transaction,
+	}
+
+	if blobberID == "" {
+		return nil, common.NewError("add_challenges",
```
do this empty checking earlier where it was acquired to avoid unnecessary works.

peterlimg commented on code/go/0chain.net/smartcontract/storagesc/models.go:
```diff
@@ -99,10 +102,10 @@ type ChallengeResponse struct {
+	BlobberID                string            `json:"blobber_id"`
+	LatestCompletedChallenge *StorageChallenge `json:"lastest_completed_challenge"`
+	ChallengeIDs             []string          `json:"challenge_ids"`
+	ChallengeIDMap           map[string]bool   `json:"-"`
```
Use struct{} instead of bool as the map value

peterlimg commented on code/go/0chain.net/smartcontract/storagesc/models.go:
```diff
@@ -150,42 +153,109 @@ func (sn *BlobberChallenge) UnmarshalMsg(data []byte) ([]byte, error) {
+}
+
+func (sn *AllocationChallenge) GetKey(globalKey string) datastore.Key {
+	return datastore.Key(globalKey + ":allocationchallenge:" + sn.AllocationID)
```
Unnecessary datastore.Key conversion as it's an alias of type `string`.

peterlimg commented on code/go/0chain.net/smartcontract/storagesc/models.go:
```diff
@@ -442,8 +537,9 @@ type StorageAllocationStats struct {
+	BlobberID    string `json:"blobber_id"`
+	AllocationID string `json:"allocation_id"`
+	// Size is blobber allocation maximum size
 	Size            int64                   `json:"size"`
```
If it means the maximum size, we'd better rename it to `Cap` or `MaxSize`, etc. Maybe change it in the future

ssardana96 committed:
Merge branch 'staging' into feat/generate_challenge

ssardana08 committed:
- pr review comments

ssardana96 commented:
Green check pasts: https://github.com/0chain/0chain/runs/5819707009?check_suite_focus=true

ssardana08 committed:
- removed for loop

peterlimg reviewed (approved):
lgtm

peterlimg committed:
Merge branch 'staging' into feat/generate_challenge
